### ISSUE TITLE
Incorrect _valueChangeOf_ payload if the publication is called by autoSet

### ISSUE THREAD

**Issue by aviriel** [Alfresco/Aikau#1049]
It's me again. I tried to find a hack for #1047 and decided to create new hidden control on the form, the value of which will be updated automatically depending on values of two other controls. So I can make a visibilityConfig for a label on the form depending on this one control. 

VisibilityConfig for a label "works" if I set the value of PRESET manually (it's a textbox), but if I change values of Select fields, then value of PRESET is updating, but label visibility is not changing. Looking at PubSub log, I can say that if the value is set by autoSetConfig, then _valueChangeOf payload does not contain "value" field. How can it happen and what is the way to solve it?

Automatically set value:
![automatically set value](https://cloud.githubusercontent.com/assets/596948/15145715/e8b8d214-16bf-11e6-88a1-c523520e16c2.png)

Manually set value:
![manually set value](https://cloud.githubusercontent.com/assets/596948/15145718/ebe267f2-16bf-11e6-9202-756dc9e221b3.png)

Form:
```
    {
    	name: "alfresco/forms/DynamicForm",
    	config: {
    		publishValueSubscriptions: ["ALF_ENTER_KEY_PRESSED"],
    		okButtonPublishTopic: "TEST",
    		showCancelButton: false,
    		widgets: [{
    			name: "alfresco/forms/controls/Select",
    			config: {
    				name: "A",
    				label: "A",
    				fieldId: "PRESET_A",
    				value: "1",
    				optionsConfig: {
    					fixed: [{
    						"label": "1",
    						"value": "1"
    					}, {
    						"label": "2",
    						"value": "2"
    					}]
    				}
    			}
    		}, {
    			name: "alfresco/forms/controls/Select",
    			config: {
    				name: "B",
    				label: "B",
    				fieldId: "PRESET_B",
    				value: "1",
    				optionsConfig: {
    					fixed: [{
    						"label": "1",
    						"value": "1"
    					}, {
    						"label": "2",
    						"value": "2"
    					}]
    				}
    			}
    		}, {
    			name: "alfresco/forms/controls/TextBox",
    			config: {
    				name: "hidden",
    				fieldId: "PRESET",
    				value: "11",
    				autoSetConfig: [{
    					rulePassValue: "11",
    					rulesMethod: "ALL",
    					rules: [{
    						targetId: "PRESET_A",
    						is: ["1"]
    					}, {
    						targetId: "PRESET_B",
    						is: ["1"]
    					}]
    				}, {
    					rulePassValue: "12",
    					rulesMethod: "ALL",
    					rules: [{
    						targetId: "PRESET_A",
    						is: ["1"]
    					}, {
    						targetId: "PRESET_B",
    						is: ["2"]
    					}]
    				}, {
    					rulePassValue: "21",
    					rulesMethod: "ALL",
    					rules: [{
    						targetId: "PRESET_A",
    						is: ["2"]
    					}, {
    						targetId: "PRESET_B",
    						is: ["1"]
    					}]
    				}, {
    					rulePassValue: "22",
    					rulesMethod: "ALL",
    					rules: [{
    						targetId: "PRESET_A",
    						is: ["2"]
    					}, {
    						targetId: "PRESET_B",
    						is: ["2"]
    					}]
    				}]
    			}
    		}, {
    			name: "alfresco/html/Label",
    			config: {
    				label: "Label",
    				visibilityConfig: {
    					initialValue: true,
    					rulesMethod: "ALL",
    					rules: [{
    						topic: "_valueChangeOf_PRESET",
    						attribute: "value",
    						is: ["11"],
    						strict: true
    					}]
    				}
    			}
    		}]
    	}
    }
```

Also I attach my project just in case (clean project generated with SDK 2.2, Aikau 1.0.67 in dependencies, this form and DebugLog are added to http://localhost:8081/share/page/dp/ws/simple-page, no other changes).

[componentX-share.zip](https://github.com/Alfresco/Aikau/files/257106/componentX-share.zip)

**Comment by AFaust** [comment]
The payload of the valueChangeOf pubSub actually contains the proper value of the widget. The issue here is that for every set of autoSetConfig rules, either rulePassValue or ruleFailValue is being set. Each of these sets is processed in the order they are defined, meaning that there will always be sets of rules evaluated after the rulePassValue "11" has been set. Since those other rule sets fail but don't define a ruleFailValue, ```undefined``` is set as the new widget value, which results in the payload not containing any value.
Confusion arises from the fact that the widget will continue to display the last non-undefined value, which is "11" in this case.
When you look in entire the pubSub log you will see that there are valueChangeOf pubSub events for every set of rules being evaluted and that the value "11" is actually contained in the very first valueChangeOf pubSub after setting the select boxes to the correct values.

**Comment by aviriel** [comment]
Tested with applied #1051 . Works now.

**Comment by draperd** [comment]
Does this mean that you're happy for this issue to be closed?

**Comment by AFaust** [comment]
I think @aviriel meant that it works with the PR and that would need to be merged / included somehow before closing.

**Comment by aviriel** [comment]
Exactly as @AFaust said. It works with the PR. It would be nice if you can merge it.

**Comment by draperd** [comment]
OK, right... just catching up with the issues.... again, there's no unit on that PR. We like to reproduce the problem with a failing unit test that we then make changes to fix and prevent future regressions. I'll raise a JIRA ticket for us to add the unit test and use @AFaust's proposed fix. However, we won't just merge without associated unit tests.

**Comment by draperd** [comment]
I've raised https://issues.alfresco.com/jira/browse/AKU-970 to provide the associated unit test to go with the fix. As this is a bug it will be prioritized for the next sprint so you can expect a fix in the 1.0.69 release (unless the PR is updated to include the necessary unit test).

**Comment by draperd** [comment]
This problem has now been fixed and is available in the snapshot builds and will be available in the 1.0.69 release due next Tuesday.

**Comment by draperd** [comment]
The fix for this is in the 1.0.69 release which is now available.
